url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/protocols/Node.java

```diff
@@ -0,0 +1,265 @@
+		System.out.println("Crypto Calls size = "+cryptoCalls.length);
+		memory = sortLong(new Memory(new MemoryRelation(), new KbitsScale()), protocols);
+		System.out.println("Memory size = "+memory.length);
+		finalSlowPhase = sortBoolean(new FinalSlowPhase(new FinalSlowPhaseRelation(), new NoScale<Boolean>()), protocols);
+		System.out.println("Final Slow Phase size = "+finalSlowPhase.length);
+		System.out.println("Sorting finished"); 
+	}
+
+	@Override
+	public DBProtocol getInstance() {
+		return this;
+	}
+
+	@Override
+	public String getAcronym() {
+		return "ToCompareWith";
+	}
+
+	@Override
+	public BigDecimal getMafiaFraudProbability() {
+		return new BigDecimal((Double)mafia[state.mafiaIndex].getValue());
+	}
+
+	@Override
+	public BigDecimal getDistanceFraudProbability() {
+		return new BigDecimal((Double)distance[state.distanceIndex].getValue());
+	}
+
+	@Override
+	public BigDecimal getTerroristFraudProbability() {
+		return new BigDecimal((Double)terrorist[state.terroristIndex].getValue());
+	}
+
+	@Override
+	public int getTotalBitsExchangedDuringFastPhase(){
+		return (Integer)bitsExchanged[state.bitsExchangedIndex].getValue();
+	}
+
+	@Override
+	public boolean hasFinalSlowPhase() {
+		return (Boolean)finalSlowPhase[state.finalSlowPhaseIndex].getValue();
+	}
+
+	@Override
+	public int getSizeOfTheChannel() {
+		return (Integer)sizeOfMessage[state.sizOfMessageIndex].getValue();
+	}
+
+	@Override
+	public final long getMemory(){
+		return (Long)memory[state.memoryIndex].getValue();
+	}
+	@Override
+	public int getCryptoCalls() {
+		return (Integer)cryptoCalls[state.cryptoCallsIndex].getValue();
+	}
+
+	@Override
+	public boolean lackSecurityProof() {
+		return false;
+	}
+
+	@Override
+	public int getYearOfPublication() {
+		return 0;
+	}
+
+	@Override
+	public int getTotalMsgSizeReceived() {
+		return 0;
+	}
+
+	@Override
+	public int getTotalOutputSizeOfCallsToFunctions() {
+		return 0;
+	}
+
+	@Override
+	public int getNumberOfNoncesGenerated() {
+		return 0;
+	}
+
+	@Override
+	public int getBitsGenerated() {
+		return 0;
+	}
+
+	@Override
+	public DBProtocol[] getInstances() {
+		return null;
+	}
+
+	@Override
+	public String getIdentifier() {
+		return "ToCompareWith";
+	}
+
+	
+	private static Attribute[] sortProbability(
+			Attribute<Double> probability, DBProtocol[] protocols) {
+		TreeMap<Double, Attribute<Double>> result = new TreeMap<>();
+		for (DBProtocol p : protocols) {
+			Attribute<Double> att = p.getAttribute(probability);
+			result.put(att.getValue(), att);
+		}
+		Attribute<Double>[] array = new Attribute[result.size()];
+		int index = 0;
+		for (Attribute<Double> a : result.values()){
+			if (index > 0 && probability.getEquality().equal(a.getValue(), array[index-1].getValue())) continue;
+			array[index] = a;
+			index++;
+		}
+		Attribute[] resultFinal = new Attribute[index];
+		System.arraycopy(array, 0, resultFinal, 0, index);
+		return resultFinal;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 152:166, 171:185, 191:205.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.